docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,140 @@
+
+## Introduction
+
+The MGC has the ability to extend the DNS configuration of the gateway with the DNSPolicy resource. This resource allows 
+users to configure health checks. As a result of configuring health checks, the controller creates the health checks in 
+Route53, attaching them to the related DNS records. This has the benefit of automatically disabling an endpoint if it 
+becomes unhealthy, and enabling it again when it becomes healthy again.
+
+This feature has a few shortfalls:
+1. It’s tightly coupled with Route53. If other DNS providers are supported they must either provide a similar feature, 
+or health checks will not be supported
+2. Lacks the ability to reach endpoints in private networks
+3. requires using the gateway controller to implement, maintain and test multiple providers
+
+This document describes a proposal to extend the current health check implementation to overcome these shortfalls.
+
+### Goals
+
+* Ability to configure health checks in the DNSPolicy associated to a Gateway
+* DNS records are disabled when the associated health check fails
+* Current status of the defined health checks is visible to the end user
+
+### Nongoals
+* Ability for the health checks to reach endpoints in private networks
+* Transparently keep support for other health check providers like Route53
+
+## Proposal
+
+Currently, this functionality will be added to the existing MGC, and executed within that component. This will be created
+with the knowledge that it may need to be made into an external component in the future.
+
+#### `DNSPolicy` resource
+
+The presence of the `healthCheck` means that for every DNS endpoint (that is either an A record, or a CNAME to an external host), 
+a health check is created based on the health check configuration in the DNSPolicy.
+
+A `failureThreshold` field will be added to the health spec, allowing users to configure a number of consecutive health 
+check failures that must be observed before the endpoint is considered unhealthy.
+
+Example DNS Policy with a defined health check.
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSPolicy
+metadata:
+  name: prod-web
+  namespace: multi-cluster-gateways
+spec:
+  healthCheck:
+    endpoint: /health
+    failureThreshold: 5
+    port: 443
+    protocol: https
+    additionalHeaders:
+      - name: "..."
+        value: "..."
+      - name: "..."
+        value: "..."
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: Gateway
+    name: prod-web
+    namespace: multi-cluster-gateways
+```
+#### `DNSHealthCheckProbe` resource
+
+The DNSHealthCheckProbe resource configures a health probe in the controller to perform the health checks against an 
+identified final A or CNAME endpoint. When created by the controller as a result of a DNS Policy, this will have an 
+owner ref of the DNS Policy that caused it to be created.
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSHealthCheckProbe
+metadata:
+  name: example-probe
+spec:
```

Comment:
Some of the fields here look similar to [kubernetes readiness & liveness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/).
Is that the inspiration for these fields, or some other standard or service?
e.g. https://docs.datadoghq.com/integrations/dns_check/ or https://documentation.solarwinds.com/en/success_center/pingdom/content/topics/what-is-a-check-.htm

Reply:
@sergioifg94 wrote the original spec for these, I am not sure where the inspiration came from.